windows/src/engine/keyman32/DebugEventTrace.cpp

```diff
@@ -0,0 +1,128 @@
+#include "keyman-debug-etw.h"
+
+#ifdef _WIN64
+#define DEBUG_PLATFORM_STRING "x64"
+#define DEBUG_PLATFORM_STRINGW L"x64"
+#else
+#define DEBUG_PLATFORM_STRING "x86"
+#define DEBUG_PLATFORM_STRINGW L"x86"
+#endif
+
+#define TAB L"\t"
+
+
+int GetActualShiftState() {   // I4843
+  int state = 0;
+  if (GetKeyState(VK_NUMLOCK) & 1) state |= NUMLOCKFLAG;
+  if (GetKeyState(VK_CAPITAL) & 1) state |= CAPITALFLAG;
+  if (GetKeyState(VK_SHIFT) < 0) state |= K_SHIFTFLAG;
+  if (GetKeyState(VK_RCONTROL) < 0) state |= RCTRLFLAG;
+  if (GetKeyState(VK_LCONTROL) < 0) state |= LCTRLFLAG;
+  if (GetKeyState(VK_LMENU) < 0) state |= LALTFLAG;
+  if (GetKeyState(VK_RMENU) < 0) state |= RALTFLAG;
+
+  if (GetKeyState(VK_LSHIFT) < 0) state |= 0x10000;
+  if (GetKeyState(VK_RSHIFT) < 0) state |= 0x20000;
+
+  return state;
+}
+
+extern "C" void _declspec(dllexport) WINAPI Keyman_WriteDebugEvent(char *file, int line, PWCHAR msg) {
+  if (file == NULL || msg == NULL) {
+    return;
+  }
+
+  PKEYMAN64THREADDATA _td = ThreadGlobals();
+  if (!_td) {
+    return;
+  }
+
+  if (!Globals::get_debug_KeymanLog()) {
+    return;
+  }
+
+  GUITHREADINFO gti;
+  WCHAR sProcessPath[256], sProcessName[32];
+
+  gti.cbSize = sizeof(gti);
+  GetGUIThreadInfo(GetCurrentThreadId(), &gti);
+
+  GetModuleFileNameW(NULL, sProcessPath, 256);
+  _wsplitpath_s(sProcessPath, NULL, 0, NULL, 0, sProcessName, 32, NULL, 0);
+  sProcessName[31] = 0;
+
+  if (wcslen(msg) > 256) {
+    msg[255] = 0;
+  }
+
+  if (Globals::get_debug_ToConsole()) {   // I3951
+    WCHAR windowinfo[1024];
+    wsprintfW(windowinfo,
+      DEBUG_PLATFORM_STRINGW TAB //"Platform" TAB
+      L"%s" TAB  //"Process" TAB
+      L"%x" TAB  //"PID" TAB
+      L"%x" TAB  //"TID" TAB
+      L"%x" TAB  //"ShiftState" TAB
+      L"%x" TAB  //"ActualShiftState" TAB
+      L"%d" TAB  //"TickCount" TAB
+      L"%x" TAB  //"FocusHWND" TAB
+      L"%8x" TAB //"ActiveHKL" TAB
+      L"%hs:%d" TAB  //"SourceFile" TAB
+      L"%s\n",     //"Message"
+      sProcessName,                    //"Process" TAB
+      GetCurrentProcessId(),           //"PID" TAB
+      GetCurrentThreadId(),            //"TID" TAB
+      Globals::get_ShiftState(),       //"ShiftState" TAB
+      GetActualShiftState(),           // ActualShiftState TAB
+      GetTickCount(),                  //"TickCount" TAB
+      gti.hwndFocus,                   //"FocusHWND" TAB
+      GetKeyboardLayout(0),            //"ActiveHKL" TAB
+      file, line,                      //"SourceFile" TAB
+      msg);                            //"Message"
+
+    OutputDebugStringW(windowinfo);
+  }
+
+  if (_td->etwRegHandle != NULL) {
+#define MAX_DESCRIPTORS 12
+    EVENT_DATA_DESCRIPTOR Descriptors[MAX_DESCRIPTORS];
+    DWORD pid = GetCurrentProcessId();
+    DWORD tid = GetCurrentThreadId();
+    DWORD shiftState = Globals::get_ShiftState();
+    DWORD actualShiftState = GetActualShiftState();
+    DWORD tickCount = GetTickCount();
+    HKL activeHKL = GetKeyboardLayout(0);
```

Comment:
```
    DWORD pid = GetCurrentProcessId();
    DWORD tid = GetCurrentThreadId();
    DWORD shiftState = Globals::get_ShiftState();
    DWORD actualShiftState = GetActualShiftState();
    DWORD tickCount = GetTickCount();
    HKL activeHKL = GetKeyboardLayout(0);
```
Can this block get moved above line 60? Then you don't need to repeat the calls in ll. 75-81

Reply:
OK